Sources/DataSource.swift

```diff
@@ -0,0 +1,126 @@
+//
+//  CellKitDataSource.swift
+//  CellKit-iOS
+//
+//  Created by Petr Zvoníček on 22.06.18.
+//  Copyright © 2018 FUNTASTY Digital, s.r.o. All rights reserved.
+//
+
+import UIKit
+
+public protocol DataSource {
+    associatedtype Section
+
+    var sections: [Section] { get }
+    var delegate: CellModelDataSourceDelegate? { get }
+    var first: Section? { get }
+
+    subscript(index: Int) -> Section { get }
+}
+
+public class AbstractDataSource: NSObject {
```

Comment:
Is there any really good reason to have this abstract class instead of protocol? For public API I argue, that this could easily be used incorrectly, the API does not tell the developer how to use it and the developer does not know about the usage at compile time. I suggest we either:

- Make `AbstractDataSource` protocol instead of class if possible
- Make this API internal

Reply:
Unfortunately there is a reason. UITableViewDataSource and UITableViewDelegate cannot be implemented as extension on Swift protocol. Making it internal is reasonable, there should be no problem with that.

Reply:
All right, so `AbstractDataSource` cannot be internal either 😞 

![obrazek](https://user-images.githubusercontent.com/1540564/41844484-a86dfdf0-7870-11e8-93da-4fa5d9a70f64.png)

Reply:
Of course we can get rid of `AbstractDataSource` entirely, for the cost of duplicating all the Delegate and DataSource conformances.

Reply:
The closest I can get to what you proposed is this

```swift
public class AbstractDataSource: NSObject {

    internal override init() { }

    // ...
}
```

Reply:
After I have given the subject some thought I think we should not separate the implementation to extensions. When the protocol inherits from the protocols I looks the best from my point of view. It is still not ideal, but we do not have the solution that would be the best.

I am thinking about support for different platforms though and in this case we would need to implement another protocol, since in this case we have something like `UIDataSource`.

```swift
public protocol AbstractDataSource: UITableViewDataSource, UICollectionViewDataSource, UITableViewDelegate, UICollectionViewDelegate {
    var delegate: CellModelDataSourceDelegate? { get set }

    func numberOfSections() -> Int
    func cellModels(in section: Int) -> [CellModel]
    func header(in section: Int) -> SupplementaryViewModel?
    func footer(in section: Int) -> SupplementaryViewModel?
    func cellModel(at indexPath: IndexPath) -> CellModel
}
```

Reply:
As we discussed in person I would suggest, that we use this variant until we find out better solution or Swift itself allows us to improve it.